block-storage/src/main/scala/coop/rchain/blockstorage/BlockStore.scala

```diff
@@ -1,31 +1,40 @@
 package coop.rchain.blockstorage
 
-import cats.Applicative
+import cats.effect.Sync
 import cats.syntax.all._
-import coop.rchain.casper.protocol.{ApprovedBlock, BlockMessage}
+import coop.rchain.blockstorage.dag.codecs.{codecBlockHash, codecBlockMessage}
+import coop.rchain.casper.protocol.{BlockMessage, BlockMessageProto}
 import coop.rchain.models.BlockHash.BlockHash
-
-trait BlockStore[F[_]] {
-  def get(blockHash: BlockHash): F[Option[BlockMessage]]
-
-  def put(f: => (BlockHash, BlockMessage)): F[Unit]
-
-  def contains(blockHash: BlockHash)(implicit applicativeF: Applicative[F]): F[Boolean] =
-    get(blockHash).map(_.isDefined)
-
-  def getApprovedBlock: F[Option[ApprovedBlock]]
-
-  def putApprovedBlock(block: ApprovedBlock): F[Unit]
-
-  // Defaults
-
-  def put(blockMessage: BlockMessage): F[Unit] =
-    put((blockMessage.blockHash, blockMessage))
-
-  def put(blockHash: BlockHash, blockMessage: BlockMessage): F[Unit] =
-    put((blockHash, blockMessage))
-}
-
-object BlockStore {
-  def apply[F[_]](implicit instance: BlockStore[F]): BlockStore[F] = instance
+import coop.rchain.shared.Compression
+import coop.rchain.shared.syntax._
+import coop.rchain.store.{KeyValueStoreManager, KeyValueTypedStore}
+import net.jpountz.lz4.{LZ4CompressorWithLength, LZ4DecompressorWithLength}
+
+object blockStore {
+  type BlockStore[F[_]] = KeyValueTypedStore[F, BlockHash, BlockMessage]
+  def BlockStore[F[_]](implicit instance: BlockStore[F]): instance.type = instance
+
+  def create[F[_]: Sync](
+      kvm: KeyValueStoreManager[F]
+  ): F[KeyValueTypedStore[F, BlockHash, BlockMessage]] =
+    kvm
+      .store("blocks")
+      .map(_.toTypedStore[BlockHash, BlockMessage](codecBlockHash, codecBlockMessage))
+
+  def bytesToBlockMessage(bytes: Array[Byte]): BlockMessage =
+    BlockMessage
+      .from(BlockMessageProto.parseFrom(decompressBytes(bytes)))
+      .getOrElse(BlockMessage.empty)
```

Comment:
`bytesToBlockMessage` should return Either directly so error can be used in scodec definition.